### ISSUE TITLE
completion_queue_threading_test is flaky (GRPC_POLL_STRATEGY=epollex)

### ISSUE THREAD

**Issue by jtattermusch** [grpc/grpc#14066]
Failing 3-4% of builds.

```
...
I0115 23:27:15.868564964    6745 completion_queue_threading_test.cc:198] consumer 106 phase 2 done
I0115 23:27:15.868603191    6754 completion_queue_threading_test.cc:198] consumer 112 phase 2 done
I0115 23:27:15.868564523    6743 completion_queue_threading_test.cc:198] consumer 104 phase 2 done
E0115 23:27:25.868801879    6330 completion_queue_threading_test.cc:261] assertion failed: gpr_event_wait(&options[i].on_finished, ten_seconds_time())
```

Seeing failures in:
dbg
https://sponge.corp.google.com/target?id=0ef9db45-f533-4c10-835e-79ac8ffc766d&target=github/grpc/c_linux_dbg_native&searchFor=&show=FAILED&sortBy=STATUS
https://sponge.corp.google.com/target?id=0a81c3e8-f823-4cfc-b106-b8b5ba3326ca&target=github/grpc/c_linux_dbg_native&searchFor=&show=FAILED&sortBy=STATUS
https://sponge.corp.google.com/target?id=c46b2c97-27f3-46a9-afe7-d49f1e894646&target=github/grpc/c_linux_dbg_native&searchFor=&show=ALL&sortBy=STATUS

asan
https://sponge.corp.google.com/target?id=732e6887-7fc8-4b98-9a84-db571cecbffb&target=github/grpc/c_linux_asan_native&searchFor=&show=ALL&sortBy=STATUS
https://sponge.corp.google.com/invocation?id=b311594a-c8c3-4d1f-a981-0ffff0c23579
https://sponge.corp.google.com/invocation?id=47254cbb-2cbd-4160-a6f4-72ccfc8e0ede

tsan
https://sponge.corp.google.com/invocation?id=8f9233b8-e387-4265-a4e0-bedbdcbda1ca
https://sponge.corp.google.com/invocation?id=a16f21b4-385a-4f45-81d6-ec616da6428c

**Comment by dgquintas** [comment]
Not sure who the best person is for this issue. Assigning to @sreecha for triage.

**Comment by jtattermusch** [comment]
I can reproduce on a VM (requires epollex)
tools/run_tests/run_tests.py -l c -c dbg -r completion_queue_threading_test -n 200 --force_use_pollers epollex

it can be reproduced with e.g `completion_queue_threading_test.cc:218] test_threading: 1 producers, 10 consumers` scenario: the producer thread seems to finish successfully, but some of the consumer threads occasionally get stuck waiting in cq_next() and they never wake up.

**Comment by jtattermusch** [comment]
I can reproduce also with test_threading: 1 producers, 2 consumers.
All the 10000 ops finalized by the producer are actually consumed correctly by the consumers (sum(events_triggered) = 10000), but the next cq_next call might never return.

**Comment by jtattermusch** [comment]
I have a good confidence I understand what's going on:

Thread 1. completion_queue_shutdown (invoked concurrently with the producers and consumers) invokes
pollset_shutdown (specific for epollex), which sets  `pollset->shutdown_closure`:
https://github.com/grpc/grpc/blob/0868da366233eb576115e2f14c0312a3080d2b85/src/core/lib/iomgr/ev_epollex_linux.cc#L737, then does pollset_kick_all to wake up everyone sleeping.

Thread 2. the consumer thread essentially runs cq_next in a loop, until SHUTDOWN event is returned. When the timing is right, `pollset_work` is invoked after the pollset_shutdown has been initiated and the `pollset->shutdown_closure` has been set - and the pollset_kick_all has already happened).
To check if `epoll_wait` should be run, it checks  ` pollset->shutdown_closure == nullptr` - which has already been set on another thread, but access to `pollset->shutdown_closure` is not synchronized.
Because `pollset->shutdown_closure` seems unset, epoll_wait with infinite deadline is run (and consequently one of the consumer threads never finishes and the test fails).
https://github.com/grpc/grpc/blob/0868da366233eb576115e2f14c0312a3080d2b85/src/core/lib/iomgr/ev_epollex_linux.cc#L865

**Comment by jtattermusch** [comment]
@sreecha  what would be the best approach for synchronizing access to  pollset->shutdown_closure.  I could either use atomics or lock with mutex (but I'm not sure which mutext is the right one ;-) )

**Comment by jtattermusch** [comment]
It seems I was wrong. Access to pollset->shutdown_closure seems to be synchronized correctly. The real rootcause:
In when the shutdown is finished in `pollset_maybe_finish_shutdown`,   `pollset->shutdown_closure` is set back to nullptr.
https://github.com/grpc/grpc/blob/master/src/core/lib/iomgr/ev_epollex_linux.cc#L564

That can still result in a never-to-wake up poll in `pollset_work` because of this line:
https://github.com/grpc/grpc/blob/0868da366233eb576115e2f14c0312a3080d2b85/src/core/lib/iomgr/ev_epollex_linux.cc#L865.

Tentative fix is #14177 (it passed 10000 runs with no failure).
